lib/src/signed_video_h26x_auth.c

```diff
@@ -353,6 +459,82 @@ set_validation_status_of_items_used_in_gop_hash(h26x_nalu_list_t *nalu_list, cha
   return num_marked_items;
 }
 
+/**
+ * Verifies the integrity of the GOP hash in the video, ensuring that the data
+ * within the GOP is authentic and complete. Updates the expected and received
+ * NALU counts, and returns true if the verification is successful.
+ *
+ * The function performs the following steps:
+ * 1. Determines the validation status based on the verified signature hash.If this signature
+ *    is not successfully verified, the entire GOP isconsidered invalid and cannot be trusted.
+ * 2. If the SEI signature is valid, the next step is to verify the GOP
+ *    hash. This hash is computed during signing and included in the SEI. On the validation side,
+ * the received GOP hash is compared with the locally computed GOP hash. If they match, the entire
+ * GOP is confirmed as valid.
+ * 3. If the GOP hash verification fails, the function attempts to
+ *    validate the GOP using individual NAL Unit hashes, provided they are available in the SEI.
+ * This secondary validation can still result in a valid GOP, even if some NAL Units are missing.
+ * 4.  Each NAL Unit in the GOP is marked according to its validation
+ *    status (valid, invalid, or missing). If necessary, missing NAL Units are added, and validation
+ *    statistics are updated to reflect the total number of expected and received NAL Units.
+ */
+static bool
+verify_hashes_with_sei(signed_video_t *self, int *num_expected_nalus, int *num_received_nalus)
+{
+  assert(self);
+
+  int num_expected_hashes = -1;
+  int num_received_hashes = -1;
+  char validation_status = 'P';
+  svrc_t status = SV_UNKNOWN_FAILURE;
+
+  // The verified_signature_hash indicates if the signature is verified.
+  // If the signature hash is verified, the GOP hash can be verified as well.
+  // If the signature hash is not verified, it means the SEI is corrupted, and the whole GOP status
+  // is determined by the verified_signature_hash.
+  if (self->gop_info->verified_signature_hash == 1) {
+    validation_status = '.';
+    num_expected_hashes = (int)self->gop_info->num_sent_nalus;
+    if (!verify_gop_hash(self)) {
+      // If the signature is verified but GOP hash is not, continue validation with the
+      // hash list if it is accessible.
+      if (self->gop_info->list_idx > 0) {
+        return verify_hashes_with_hash_list(self, num_expected_nalus, num_received_nalus);
+      }
+      validation_status = 'N';
+    }
+  } else if (self->gop_info->verified_signature_hash == 0) {
+    validation_status = 'N';
```

Comment:
Unfortunately there is no test case where a SEI has been tampered with. So it is not known if this path in the code actually works. We should add such a case in a separate PR before pushing this PR.

Reply:
Test added.